extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java

```diff
@@ -0,0 +1,82 @@
+package io.quarkus.infinispan.client.runtime.cache;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.commons.CacheException;
+
+import io.smallrye.mutiny.Uni;
+
+public class InfinispanGetWrapper {
+    final RemoteCache cache;
+    final Map<Object, CompletableFuture<Object>> synchronousGetLocks;
+
+    public InfinispanGetWrapper(RemoteCache cache, Map<Object, CompletableFuture<Object>> synchronousGetLocks) {
+        this.cache = cache;
+        this.synchronousGetLocks = synchronousGetLocks;
+    }
+
+    public <K, V> Uni<V> get(K key, Function<K, V> valueLoader) {
+        return Uni.createFrom().completionStage(
+                /*
+                 * Even if CompletionStage is eager, the Supplier used below guarantees that the cache value computation will be
+                 * delayed until subscription time. In other words, the cache value computation is done lazily.
+                 */
+                new Supplier<CompletionStage<V>>() {
+                    @Override
+                    public CompletionStage<V> get() {
+                        CompletionStage<Object> infinispanValue = getFromInfinispan(key, valueLoader);
+                        return cast(infinispanValue);
+                    }
+                });
+    }
+
+    private <K, V> CompletableFuture<Object> getFromInfinispan(K key, Function<K, V> valueLoader) {
+        CompletableFuture<Object> stage = new CompletableFuture<>();
+        CompletableFuture<Object> prev;
+        try {
+            if ((prev = synchronousGetLocks.putIfAbsent(key, stage)) != null) {
+                return prev;
+            }
+            cache.getAsync(key)
+                    .whenComplete((valGet, tGet) -> {
+                        if (tGet != null) {
+                            stage.completeExceptionally((Throwable) tGet);
+                        } else if (valGet != null) {
+                            stage.complete(valGet);
+                        } else {
+                            Object newValue = valueLoader.apply(key);
+                            if (newValue != null) {
+                                cache.putIfAbsentAsync(key, newValue)
+                                        .whenComplete((valPut, tPut) -> {
+                                            if (tPut != null) {
+                                                stage.completeExceptionally((Throwable) tPut);
+                                            } else {
+                                                stage.complete(valPut == null ? newValue : valPut);
+                                            }
+                                        });
+                            }
+                        }
+                        synchronousGetLocks.remove(key);
```

Comment:
This should be removed when the stage is completed. We are removing it too early here imo.

Thinking about this more, should we just confirm if we want to wait until the write completes before returning the value or return early. The latter will reduce latency for an individual request, but if you have concurrent requests it would be possible, but unlikely, for interleaving to do the calculation twice.